File: src/main/kotlin/com/kursx/parser/fb2/Annotation.java
package com.kursx.parser.fb2;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jsoup.Jsoup;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import java.io.StringWriter;
import java.util.ArrayList;

//http://www.fictionbook.org/index.php/Элемент_annotation
@SuppressWarnings("unused")
public class Annotation extends IdElement {

    private String text = "";
    protected String lang;
    protected ArrayList<Element> elements;

    public Annotation() {
    }

    Annotation(Node node) {
        super(node);
        try {
            StringWriter writer = new StringWriter();
            TransformerFactory.newInstance().newTransformer().transform(new DOMSource(node), new StreamResult(writer));
            String xml = writer.toString();
            this.text = Jsoup.parseBodyFragment(xml).body().text();
        } catch (TransformerException ignored) {
        }
        NamedNodeMap map = node.getAttributes();
        for (int index = 0; index < map.getLength(); index++) {
            Node attr = map.item(index);
            if (attr.getNodeName().equals("xml:lang")) {
                lang = attr.getNodeValue();
            }
        }
        NodeList nodeList = node.getChildNodes();
        for (int i = 0; i < nodeList.getLength(); i++) {
            Node paragraph = nodeList.item(i);
            if (elements == null) elements = new ArrayList<>();
            switch (paragraph.getNodeName()) {
                case "p":
                    elements.add(new P(paragraph));
                    break;
                case "poem":
                    elements.add(new Poem(paragraph));
                    break;
                case "cite":
                    elements.add(new Cite(paragraph));
                    break;
                case "subtitle":
                    elements.add(new Subtitle(paragraph));
                    break;
                case "empty-line":
                    elements.add(new EmptyLine());
                    break;
                case "table":
                    elements.add(new Table());
                    break;
            }
        }
    }

    @NotNull
    public ArrayList<Element> getAnnotations() {
        return elements == null ? new ArrayList<>() : elements;
    }

    @Nullable
    public String getLang() {
        return lang;
    }

    public void setLang(String lang) {
        this.lang = lang;
    }

    public ArrayList<Element> getElements() {
        return elements;
    }

    public void setElements(ArrayList<Element> elements) {
        this.elements = elements;
    }

    public String getText() {
        return text;
    }
}

File: src/main/kotlin/com/kursx/parser/fb2/Annotation.kt
package com.kursx.parser.fb2

import kotlinx.serialization.Serializable
import org.jsoup.Jsoup
import org.w3c.dom.Node
import java.io.StringWriter
import javax.xml.transform.TransformerException
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

//http://www.fictionbook.org/index.php/Элемент_annotation
@Suppress("unused")
@Serializable
open class Annotation : IdElement {
    var text = ""
        private set
    var lang: String? = null
    var elements: ArrayList<Element>? = null

    constructor()
    internal constructor(node: Node) : super(node) {
        try {
            val writer = StringWriter()
            TransformerFactory.newInstance().newTransformer().transform(DOMSource(node), StreamResult(writer))
            val xml = writer.toString()
            text = Jsoup.parseBodyFragment(xml).body().text()
        } catch (ignored: TransformerException) {
        }
        val map = node.attributes
        for (index in 0 until map.length) {
            val attr = map.item(index)
            if (attr.nodeName == "xml:lang") {
                lang = attr.nodeValue
            }
        }
        val nodeList = node.childNodes
        for (i in 0 until nodeList.length) {
            val paragraph = nodeList.item(i)
            if (elements == null) elements = ArrayList()
            when (paragraph.nodeName) {
                "p" -> elements!!.add(P(paragraph))
                "poem" -> elements!!.add(Poem(paragraph))
                "cite" -> elements!!.add(Cite(paragraph))
                "subtitle" -> elements!!.add(Subtitle(paragraph))
                "empty-line" -> elements!!.add(EmptyLine())
                "table" -> elements!!.add(Table())
            }
        }
    }

    val annotations: ArrayList<Element>
        get() = if (elements == null) ArrayList() else elements!!

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Annotation

        if (text != other.text) return false
        if (lang != other.lang) return false
        if (elements != other.elements) return false

        return true
    }

    override fun hashCode(): Int {
        var result = text.hashCode()
        result = 31 * result + (lang?.hashCode() ?: 0)
        result = 31 * result + (elements?.hashCode() ?: 0)
        return result
    }

}
